pyteal/compiler/compiler_test.py

```diff
@@ -2193,3 +2193,455 @@ def access_b4_store(magic_num: pt.abi.Uint64, *, output: pt.abi.Uint64):
 
     with pytest.raises(pt.TealInternalError):
         pt.compileTeal(program_access_b4_store_broken, pt.Mode.Application, version=6)
+
+
+def test_router_app():
+    router = pt.Router()
+
+    @router.add_method_handler
+    @pt.ABIReturnSubroutine
+    def add(a: pt.abi.Uint64, b: pt.abi.Uint64, *, output: pt.abi.Uint64) -> pt.Expr:
+        return output.set(a.get() + b.get())
+
+    @router.add_method_handler
```

Comment:
I'm not convinced that this is the best use-case for a decorator, but maybe would be after further discussions. The reason I hesitate is that I typically think of function-decorators as "side-effectless", i.e., decoratoring a function shouldn't mutate anything other than the thing being decorated (and technically it doesn't, as it just creates a new object from the original). Here, we are asking developers to create routers by using this unusual pattern. If there are other venerated examples in the Python world that do this, it would go a long way to ameliorating my concern.

Reply:
Ok. I've convinced myself that my concerns are misplaced. For example we can see that [fast api](https://fastapi.tiangolo.com/tutorial/first-steps/) follows a similar approach:
<img width="321" alt="image" src="https://user-images.githubusercontent.com/291133/169202336-e148532b-b55a-4ddc-88f0-32c9a31b4292.png">

So I'm crossing out my previous comment.

Reply:
But we should still be deliberate about our choices. 

We can use Fast API's approach as a guide. 

To declare -as in [these examples](https://fastapi.tiangolo.com/tutorial/bigger-applications/#path-operations-with-apirouter)- that `async_ def read_user_me()` has the route `users/me` and `tags=["users"]` you do the following:
```python
@router.get("/users/me", tags=["users"])
async def read_user_me():
    ...
```
So even though this is _adding_ a route with verb `get`, FastAPI uses `@get()` rather than `@add_get()`. Similarly, it makes sense that we could do something like:

```python
router = Router()

@router.ABIMethod
def add(a: pt.abi.Uint64, b: pt.abi.Uint64, *, output: pt.abi.Uint64) -> pt.Expr:
        ...
```

Reply:
`@router.ABIMethod` would combine presently what `@router.add_method_handler` and `@pt.ABIReturnSubroutine` into a single decorator. Behind the scenes, `@router.ABIMethod` would just be the composite of the current `@router.add_method_handler` and `@pt.ABIReturnSubroutine`, but I recommend renaming and privatizing `@router.add_method_handler` to something like `@router._add_method_handler`.

Reply:
good point, I am imaging some interface just as you mentioned, with a type signature as following
```python
abi_method(on_complete: EnumInt, create: bool) -> Callable[Callable, ABIReturnSubroutine]
```

essentially, the resulting syntax would be something like
```python
@router.abi_method(on_complete=OnComplete.NoOp, create=False)
def add(a: pt.abi.Uint64, b: pt.abi.Uint64, *, output: pt.abi.Uint64) -> pt.Expr:
    ...
```
which should match what you suggested.

This brings up another question, about the necessity of keeping `signature: str` field in `_add_method_call`. If we want to keep `signature`, then the `abi_method` should also keep such field for API consistency?

Reply:
I am somewhat tend to remove such `signature` field, since it was first designed based on the assumption that, a subroutine cannot infer its own method signature, and ABI-value return is extended in `Subroutine` class.

Providing such flexibility is too much flexibility for users, I suppose.